Rework dsd intake in listeners + origin PID detection for sockets

xvello opened the pull request:
### What does this PR do?

To implement origin tagging for custom metrics, the first step is to extract the PID. **Resolving the PID to a container id will be done in another PR**.

To enable that, this PR:

  - moves the UDP/UDS listening logic to a `listeners` subpackage, used by the `Server` class. The payload handling logic stays in the `Server` class as it is common
  - the UDP logic is unchanged, tests will be updated
  - the UDS logic can now use unix syscalls to get the origin's PID if `dogstatsd_origin_detection` is true. This is behind the `linux` build flag as if makes linux-only assumptions.

dsd udp benchmark reports the same performance as master (>16k packets/s on 2 cores before first packet drop)

### TODO

- [x] update tests for UDP
- [x] update docker image integration test with new udp+socket case
- [x] create tests for UDS
- [x] document the package
- [x] test build/run on Windows : builds and socket init errors out gracefully, keeping UDP OK

xvello committed:
initial StatsdListener implementation

masci reviewed (changes requested):
Left few comments, mostly nits, otherwise LGTM!

xvello committed:
address PR comments, see full desc

- s/Payload/Packet
- s/UnixListener/UDSListener/
- add expvar to both listeners
- improve logging + add "incomplete feature" warn for origin detection
- linting
- add licence headers
- oob size is computed instead of constant
- update dogstatds docker integration test

xvello committed:
add uds listening tests

xvello committed:
add udp_test.go

for now, they are redundant with server_test.go from parent
package, but ensure coverage of the udp transmission bits.

I'm keeping server_test.go's UDP usage instead of feeding through
the packetIn channel to keep testing UDP works on all metric types
although that should be a given

masci reviewed (approved):
🍰

xvello committed:
Merge branch 'master' into xvello/dsd_listeners
